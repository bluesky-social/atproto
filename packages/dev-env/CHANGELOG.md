# @atproto/dev-env

## 0.2.17

### Patch Changes

- Updated dependencies [[`45352f9b`](https://github.com/bluesky-social/atproto/commit/45352f9b6d02aa405be94e9102424d983912ca5d)]:
  - @atproto/api@0.7.0
  - @atproto/bsky@0.0.17
  - @atproto/pds@0.3.5

## 0.2.16

### Patch Changes

- Updated dependencies [[`3c0ef382`](https://github.com/bluesky-social/atproto/commit/3c0ef382c12a413cc971ae47ffb341236c545f60), [`378fc613`](https://github.com/bluesky-social/atproto/commit/378fc6132f621ca517897c9467ed5bba134b3776)]:
  - @atproto/syntax@0.1.5
  - @atproto/api@0.6.24
  - @atproto/bsky@0.0.16
  - @atproto/lexicon@0.3.1
  - @atproto/pds@0.3.4
  - @atproto/xrpc-server@0.4.2

## 0.2.15

### Patch Changes

- Updated dependencies [[`e1b5f253`](https://github.com/bluesky-social/atproto/commit/e1b5f2537a5ba4d8b951a741269b604856028ae5)]:
  - @atproto/crypto@0.3.0
  - @atproto/xrpc-server@0.4.1
  - @atproto/bsky@0.0.15
  - @atproto/identity@0.3.2
  - @atproto/pds@0.3.3
  - @atproto/api@0.6.23

## 0.2.14

### Patch Changes

- Updated dependencies [[`772736a0`](https://github.com/bluesky-social/atproto/commit/772736a01081f39504e1b19a1b3687783bb78f07)]:
  - @atproto/api@0.6.23
  - @atproto/bsky@0.0.14
  - @atproto/pds@0.3.2

## 0.2.13

### Patch Changes

- [#1788](https://github.com/bluesky-social/atproto/pull/1788) [`84e2d4d2`](https://github.com/bluesky-social/atproto/commit/84e2d4d2b6694f344d80c18672c78b650189d423) Thanks [@bnewbold](https://github.com/bnewbold)! - update license to "MIT or Apache2"

- Updated dependencies [[`ce49743d`](https://github.com/bluesky-social/atproto/commit/ce49743d7f8800d33116b88001d7b512553c2c89), [`84e2d4d2`](https://github.com/bluesky-social/atproto/commit/84e2d4d2b6694f344d80c18672c78b650189d423)]:
  - @atproto/lexicon@0.3.0
  - @atproto/xrpc-server@0.4.0
  - @atproto/common-web@0.2.3
  - @atproto/identity@0.3.1
  - @atproto/crypto@0.2.3
  - @atproto/syntax@0.1.4
  - @atproto/bsky@0.0.13
  - @atproto/api@0.6.22
  - @atproto/pds@0.3.1

## 0.2.12

### Patch Changes

- Updated dependencies [[`9c98a5ba`](https://github.com/bluesky-social/atproto/commit/9c98a5baaf503b02238a6afe4f6e2b79c5181693), [`bb039d8e`](https://github.com/bluesky-social/atproto/commit/bb039d8e4ce5b7f70c4f3e86d1327e210ef24dc3), [`35d108ce`](https://github.com/bluesky-social/atproto/commit/35d108ce94866ce1b3d147cd0620a0ba1c4ebcd7)]:
  - @atproto/api@0.6.21
  - @atproto/identity@0.3.0
  - @atproto/common-web@0.2.2
  - @atproto/bsky@0.0.12
  - @atproto/pds@0.3.0
  - @atproto/lexicon@0.2.3
  - @atproto/syntax@0.1.3
  - @atproto/xrpc-server@0.3.3

## 0.2.11

### Patch Changes

- [#1568](https://github.com/bluesky-social/atproto/pull/1568) [`41ee177f`](https://github.com/bluesky-social/atproto/commit/41ee177f5a440490280d17acd8a89bcddaffb23b) Thanks [@dholms](https://github.com/dholms)! - Added email verification and update flows

- Updated dependencies [[`41ee177f`](https://github.com/bluesky-social/atproto/commit/41ee177f5a440490280d17acd8a89bcddaffb23b), [`41ee177f`](https://github.com/bluesky-social/atproto/commit/41ee177f5a440490280d17acd8a89bcddaffb23b)]:
  - @atproto/api@0.6.20
  - @atproto/pds@0.1.20
  - @atproto/common-web@0.2.1
  - @atproto/bsky@0.0.11
  - @atproto/identity@0.2.1
  - @atproto/syntax@0.1.2
  - @atproto/xrpc-server@0.3.2

## 0.2.10

### Patch Changes

- Updated dependencies [[`35b616cd`](https://github.com/bluesky-social/atproto/commit/35b616cd82232879937afc88d3f77d20c6395276)]:
  - @atproto/api@0.6.19
  - @atproto/bsky@0.0.10
  - @atproto/pds@0.1.19

## 0.2.9

### Patch Changes

- Updated dependencies [[`2ce8a11b`](https://github.com/bluesky-social/atproto/commit/2ce8a11b8daf5d39027488c5dde8c47b0eb937bf)]:
  - @atproto/api@0.6.18
  - @atproto/bsky@0.0.9
  - @atproto/pds@0.1.18

## 0.2.8

### Patch Changes

- Updated dependencies [[`d96f7d9b`](https://github.com/bluesky-social/atproto/commit/d96f7d9b84c6fbab9711059c8584a77d892dcedd)]:
  - @atproto/bsky@0.0.8
  - @atproto/api@0.6.17
  - @atproto/pds@0.1.17

## 0.2.7

### Patch Changes

- Updated dependencies [[`56e2cf89`](https://github.com/bluesky-social/atproto/commit/56e2cf8999f6d7522529a9be8652c47545f82242)]:
  - @atproto/api@0.6.16
  - @atproto/bsky@0.0.7
  - @atproto/pds@0.1.16

## 0.2.6

### Patch Changes

- Updated dependencies [[`2cc329f2`](https://github.com/bluesky-social/atproto/commit/2cc329f26547217dd94b6bb11ee590d707cbd14f)]:
  - @atproto/api@0.6.15
  - @atproto/bsky@0.0.6
  - @atproto/pds@0.1.15

## 0.2.5

### Patch Changes

- Updated dependencies [[`b1dc3555`](https://github.com/bluesky-social/atproto/commit/b1dc355504f9f2e047093dc56682b8034518cf80)]:
  - @atproto/syntax@0.1.1
  - @atproto/api@0.6.14
  - @atproto/bsky@0.0.5
  - @atproto/pds@0.1.14
  - @atproto/xrpc-server@0.3.1

## 0.2.4

### Patch Changes

- Updated dependencies [[`3877210e`](https://github.com/bluesky-social/atproto/commit/3877210e7fb3c76dfb1a11eb9ba3f18426301d9f)]:
  - @atproto/api@0.6.13
  - @atproto/bsky@0.0.4
  - @atproto/pds@0.1.13
