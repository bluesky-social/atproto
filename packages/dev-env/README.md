# ATP Developer Environment

A command-line application for developers to construct and manage development environments.

## REPL API

The following methods are available in the REPL.

### `status()`

List the currently active servers.

### `startPds(port?: number)`

Create a new PDS instance. Data is stored in memory.

### `stop(port: number)`

Stop the server at the given port.

### `mkuser(handle: string, pdsPort?: number)`

Create a new user.

### `user(handle: string): ServiceClient`

Get the `ServiceClient` for the given user.