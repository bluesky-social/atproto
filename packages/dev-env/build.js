const { copy } = require('esbuild-plugin-copy')

require('esbuild')
  .build({
    logLevel: 'info',
    entryPoints: ['src/index.ts', 'src/cli.ts'],
    bundle: true,
    sourcemap: true,
    outdir: 'dist',
    platform: 'node',
    external: [
      'better-sqlite3',
      // Referenced in pg driver, but optional and we don't use it
      'pg-native',
      'sharp',
    ],
    plugins: [
      copy({
        assets: {
          from: ['../pds/src/mailer/templates/**/*'],
          to: ['./templates'],
          keepStructure: true,
        },
      }),
    ],
  })
  .catch(() => process.exit(1))
