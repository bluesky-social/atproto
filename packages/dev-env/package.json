{
  "name": "@atproto/dev-env",
  "version": "0.1.0",
  "main": "src/test-env.ts",
  "bin": "dist/cli.js",
  "license": "MIT",
  "repository": "git@github.com:bluesky-social/atproto.git",
  "scripts": {
    "build": "node ./build.js",
    "postbuild": "tsc --build tsconfig.build.json",
    "start": "node dist/cli.js",
    "prettier": "prettier --check src/",
    "prettier:fix": "prettier --write src/",
    "lint": "eslint . --ext .ts,.tsx",
    "lint:fix": "yarn lint --fix",
    "verify": "run-p prettier lint",
    "verify:fix": "yarn prettier:fix && yarn lint:fix"
  },
  "dependencies": {
    "@atproto/api": "*",
    "@atproto/bsky": "*",
    "@atproto/crypto": "*",
    "@atproto/did-resolver": "*",
    "@atproto/pds": "*",
    "@atproto/uri": "*",
    "@did-plc/lib": "^0.0.1",
    "@did-plc/server": "^0.0.1",
    "better-sqlite3": "^7.6.2",
    "chalk": "^5.0.1",
    "dotenv": "^16.0.1",
    "get-port": "^6.1.2",
    "sharp": "^0.31.2"
  },
  "devDependencies": {
    "ts-node": "^10.8.1"
  }
}
