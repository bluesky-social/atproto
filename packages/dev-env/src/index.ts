import http from 'http'
import chalk from 'chalk'
import crytpo from 'crypto'
import PDSServer, {
  Database as PDSDatabase,
  MemoryBlobStore,
  ServerConfig as PDSServerConfig,
  AppContext as PDSContext,
} from '@atproto/pds'
import * as bsky from '@atproto/bsky'
import * as plc from '@did-plc/lib'
import { PlcServer, Database as PlcDatabase } from '@did-plc/server'
import * as crypto from '@atproto/crypto'
import AtpAgent from '@atproto/api'
import { ServerType, ServerConfig, StartParams, PORTS } from './types.js'
import { HOUR } from '@atproto/common'
import { mockNetworkUtilities } from './test-env'

export * from './test-env'

interface Startable {
  start(): Promise<http.Server>
}

interface Destroyable {
  destroy(): Promise<void>
}

export class DevEnvServer {
  inst?: Destroyable

  constructor(
    private env: DevEnv,
    public type: ServerType,
    public port: number,
  ) {}

  get name() {
    return {
      [ServerType.PersonalDataServer]: 'ðŸŒž Personal Data server',
      [ServerType.DidPlaceholder]: 'ðŸ‘¤ DID Placeholder server',
      [ServerType.BskyAppView]: 'ðŸŒ€ Bsky App View server',
    }[this.type]
  }

  get description() {
    return `[${chalk.bold(this.port)}] ${this.name}`
  }

  get url() {
    return `http://localhost:${this.port}`
  }

  get ctx(): PDSContext | undefined {
    if (this.inst instanceof PDSServer) {
      return this.inst.ctx
    }
  }

  async start() {
    if (this.inst) {
      throw new Error('Already started')
    }

    const startServer = async (server: Startable): Promise<void> => {
      try {
        await server.start()
        console.log(`${this.description} started ${chalk.gray(this.url)}`)
      } catch (err) {
        console.log(`${this.description} failed to start:`, err)
      }
    }

    switch (this.type) {
      case ServerType.PersonalDataServer: {
        if (!this.env.plcUrl) {
          throw new Error('Must be running a PLC server to start a PDS')
        }

        const db = await PDSDatabase.memory()
        await db.migrateToLatestOrThrow()
        const keypair = await crypto.EcdsaKeypair.create()

        const blobstore = new MemoryBlobStore()

        const plcClient = new plc.Client(this.env.plcUrl)
        const serverDid = await plcClient.createDid({
          signingKey: keypair.did(),
          rotationKeys: [keypair.did()],
          handle: 'localhost',
          pds: `http://localhost:${this.port}`,
          signer: keypair,
        })

        const pds = PDSServer.create({
          db,
          blobstore,
          repoSigningKey: keypair,
          plcRotationKey: keypair,
          config: new PDSServerConfig({
            debugMode: true,
            version: '0.0.0',
            scheme: 'http',
            hostname: 'localhost',
            port: this.port,
            didPlcUrl: this.env.plcUrl,
            serverDid,
            recoveryKey: keypair.did(),
            jwtSecret: crytpo.randomBytes(8).toString('base64'),
            availableUserDomains: ['.test'],
            appUrlPasswordReset: 'app://password-reset',
            // @TODO setup ethereal.email creds and set emailSmtpUrl here
            emailNoReplyAddress: 'noreply@blueskyweb.xyz',
            adminPassword: 'password',
            inviteRequired: false,
            userInviteInterval: null,
            imgUriSalt: '9dd04221f5755bce5f55f47464c27e1e',
            imgUriKey:
              'f23ecd142835025f42c3db2cf25dd813956c178392760256211f9d315f8ab4d8',
            privacyPolicyUrl: 'https://example.com/privacy',
            termsOfServiceUrl: 'https://example.com/tos',
            labelerDid: 'did:example:labeler',
            labelerKeywords: {},
            maxSubscriptionBuffer: 200,
            repoBackfillLimitMs: HOUR,
            appViewRepoProvider: process.env.APP_VIEW_REPO_PROVIDER,
          }),
        })
        await startServer(pds)
        this.inst = pds
        break
      }
      case ServerType.DidPlaceholder: {
        const db = PlcDatabase.mock()
        const plcServer = PlcServer.create({ db, port: this.port })
        await startServer(plcServer)
        this.inst = plcServer
        break
      }
      case ServerType.BskyAppView: {
        if (!this.env.pdsUrl || !this.env.plcUrl) {
          throw new Error(
            'Must be running a PDS and PLC servers to start AppView',
          )
        }

        const config = new bsky.ServerConfig({
          version: '0.0.0',
          didPlcUrl: this.env.plcUrl,
          publicUrl: 'https://bsky.public.url',
          imgUriSalt: '9dd04221f5755bce5f55f47464c27e1e',
          imgUriKey:
            'f23ecd142835025f42c3db2cf25dd813956c178392760256211f9d315f8ab4d8',
          adminPassword: 'password',
          labelerDid: 'did:example:labeler',
          dbPostgresUrl: process.env.DB_POSTGRES_URL || '',
          dbPostgresSchema: process.env.DB_POSTGRES_SCHEMA,
          repoProvider: this.env.pdsUrl.replace('http://', 'ws://'),
          port: this.port,
          labelerKeywords: {},
        })

        const db = bsky.Database.postgres({
          url: config.dbPostgresUrl,
          schema: config.dbPostgresSchema,
        })

        await db.migrateToLatestOrThrow()

        const server = bsky.BskyAppView.create({ db, config })
        await startServer(server)
        this.inst = server
        break
      }
      default:
        throw new Error(`Unsupported server type: ${this.type}`)
    }
  }

  async close() {
    if (this.inst) {
      console.log(`Closing ${this.description}`)
      await this.inst.destroy()
    }
  }

  getClient(): AtpAgent {
    return new AtpAgent({ service: `http://localhost:${this.port}` })
  }
}

export class DevEnv {
  plcUrl: string | undefined
  pdsUrl: string | undefined
  servers: Map<number, DevEnvServer> = new Map()

  static async create(params: StartParams): Promise<DevEnv> {
    const devEnv = new DevEnv()
    for (const cfg of params.servers || []) {
      await devEnv.add(cfg)
    }
    return devEnv
  }

  async add(cfg: ServerConfig) {
    if (this.servers.has(cfg.port)) {
      throw new Error(`Port ${cfg.port} is in use`)
    } else if (cfg.type === ServerType.DidPlaceholder && this.plcUrl) {
      throw new Error('There should only be one plc server')
    }
    const server = new DevEnvServer(this, cfg.type, cfg.port)
    await server.start()
    this.servers.set(cfg.port, server)
    if (cfg.type === ServerType.DidPlaceholder) {
      this.plcUrl = `http://localhost:${cfg.port}`
    }
    if (cfg.type === ServerType.PersonalDataServer) {
      this.pdsUrl = `http://localhost:${cfg.port}`
    }
    if (cfg.type === ServerType.BskyAppView) {
      const pds = this.servers.get(PORTS[ServerType.PersonalDataServer])
      if (pds) {
        await mockNetworkUtilities({
          port: pds.port,
          url: pds.url,
          close: pds.close,
          ctx: pds.ctx!,
        })
      }
    }
  }

  async remove(server: number | DevEnvServer) {
    const port = typeof server === 'number' ? server : server.port
    const inst = this.servers.get(port)
    if (inst) {
      await inst.close()
      this.servers.delete(port)
    }
  }

  async shutdown() {
    for (const server of this.servers.values()) {
      await server.close()
    }
  }

  hasType(type: ServerType) {
    for (const s of this.servers.values()) {
      if (s.type === type) {
        return true
      }
    }
    return false
  }

  listOfType(type: ServerType): DevEnvServer[] {
    return Array.from(this.servers.values()).filter((s) => s.type === type)
  }
}
