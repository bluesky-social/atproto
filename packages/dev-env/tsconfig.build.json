{
  "extends": "../../tsconfig/node.json",
  "compilerOptions": {
    "rootDir": "./src",
    "outDir": "./dist",
    "noPropertyAccessFromIndexSignature": false
  },
  "include": ["./src"]
}
