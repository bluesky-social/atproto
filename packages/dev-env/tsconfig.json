{
  "extends": "../../tsconfig.json",
  "compilerOptions": {
    "outDir": "./dist", // Your outDir,
    "emitDeclarationOnly": true
  },
  "module": "esnext",
  "include": ["./src","__tests__/**/**.ts"],
  "references": [
    { "path": "../api/tsconfig.build.json" },
    { "path": "../bsky/tsconfig.build.json" },
    { "path": "../crypto/tsconfig.build.json" },
    { "path": "../did-resolver/tsconfig.build.json" },
    { "path": "../pds/tsconfig.json" },
    { "path": "../uri/tsconfig.build.json" },
  ]
}