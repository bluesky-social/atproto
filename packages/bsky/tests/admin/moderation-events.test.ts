import { TestNetwork, SeedClient } from '@atproto/dev-env'
import AtpAgent, { ComAtprotoAdminDefs } from '@atproto/api'
import { forSnapshot } from '../_util'
import basicSeed from '../seeds/basic'
import {
  REASONMISLEADING,
  REASONSPAM,
} from '../../src/lexicon/types/com/atproto/moderation/defs'

describe('moderation-events', () => {
  let network: TestNetwork
  let agent: AtpAgent
  let pdsAgent: AtpAgent
  let sc: SeedClient

  const emitModerationEvent = async (eventData) => {
    return pdsAgent.api.com.atproto.admin.emitModerationEvent(eventData, {
      encoding: 'application/json',
      headers: network.bsky.adminAuthHeaders('moderator'),
    })
  }

  const queryModerationEvents = (eventQuery) =>
    agent.api.com.atproto.admin.queryModerationEvents(eventQuery, {
      headers: network.bsky.adminAuthHeaders('moderator'),
    })

  const seedEvents = async () => {
    const bobsAccount = {
      $type: 'com.atproto.admin.defs#repoRef',
      did: sc.dids.bob,
    }
    const alicesAccount = {
      $type: 'com.atproto.admin.defs#repoRef',
      did: sc.dids.alice,
    }
    const bobsPost = {
      $type: 'com.atproto.repo.strongRef',
      uri: sc.posts[sc.dids.bob][0].ref.uriStr,
      cid: sc.posts[sc.dids.bob][0].ref.cidStr,
    }
    const alicesPost = {
      $type: 'com.atproto.repo.strongRef',
      uri: sc.posts[sc.dids.alice][0].ref.uriStr,
      cid: sc.posts[sc.dids.alice][0].ref.cidStr,
    }

    for (let i = 0; i < 4; i++) {
      await emitModerationEvent({
        event: {
          $type: 'com.atproto.admin.defs#modEventReport',
          reportType: i % 2 ? REASONSPAM : REASONMISLEADING,
          comment: 'X',
        },
        //   Report bob's account by alice and vice versa
        subject: i % 2 ? bobsAccount : alicesAccount,
        createdBy: i % 2 ? sc.dids.alice : sc.dids.bob,
      })
      await emitModerationEvent({
        event: {
          $type: 'com.atproto.admin.defs#modEventReport',
          reportType: REASONSPAM,
          comment: 'X',
        },
        //   Report bob's post by alice and vice versa
        subject: i % 2 ? bobsPost : alicesPost,
        createdBy: i % 2 ? sc.dids.alice : sc.dids.bob,
      })
    }
  }

  beforeAll(async () => {
    network = await TestNetwork.create({
      dbPostgresSchema: 'bsky_moderation_events',
    })
    agent = network.bsky.getClient()
    pdsAgent = network.pds.getClient()
    sc = network.getSeedClient()
    await basicSeed(sc)
    await network.processAll()
    await seedEvents()
  })

  afterAll(async () => {
    await network.close()
  })

  describe('query events', () => {
    it('returns all events for record or repo', async () => {
      const [bobsEvents, alicesPostEvents] = await Promise.all([
        queryModerationEvents({
          subject: sc.dids.bob,
        }),
        queryModerationEvents({
          subject: sc.posts[sc.dids.alice][0].ref.uriStr,
        }),
      ])

      expect(forSnapshot(bobsEvents.data.events)).toMatchSnapshot()
      expect(forSnapshot(alicesPostEvents.data.events)).toMatchSnapshot()
    })

    it('filters events by types', async () => {
      const alicesAccount = {
        $type: 'com.atproto.admin.defs#repoRef',
        did: sc.dids.alice,
      }
      await Promise.all([
        emitModerationEvent({
          event: {
            $type: 'com.atproto.admin.defs#modEventComment',
            comment: 'X',
          },
          subject: alicesAccount,
          createdBy: 'did:plc:moderator',
        }),
        emitModerationEvent({
          event: {
            $type: 'com.atproto.admin.defs#modEventEscalate',
            comment: 'X',
          },
          subject: alicesAccount,
          createdBy: 'did:plc:moderator',
        }),
      ])
      const [allEvents, reportEvents] = await Promise.all([
        queryModerationEvents({
          subject: sc.dids.alice,
        }),
        queryModerationEvents({
          subject: sc.dids.alice,
          types: ['com.atproto.admin.defs#modEventReport'],
        }),
      ])

      expect(allEvents.data.events.length).toBeGreaterThan(
        reportEvents.data.events.length,
      )
      expect(
        [...new Set(reportEvents.data.events.map((e) => e.event.$type))].length,
      ).toEqual(1)

      expect(
        [...new Set(allEvents.data.events.map((e) => e.event.$type))].length,
      ).toEqual(3)
    })

    it('returns events for all content by user', async () => {
      const [forAccount, forPost] = await Promise.all([
        queryModerationEvents({
          subject: sc.dids.bob,
          includeAllUserRecords: true,
        }),
        queryModerationEvents({
          subject: sc.posts[sc.dids.bob][0].ref.uriStr,
          includeAllUserRecords: true,
        }),
      ])

      expect(forAccount.data.events.length).toEqual(forPost.data.events.length)
      // Save events are returned from both requests
      expect(forPost.data.events.map(({ id }) => id).sort()).toEqual(
        forAccount.data.events.map(({ id }) => id).sort(),
      )
    })

    it('returns paginated list of events with cursor', async () => {
      const allEvents = await queryModerationEvents({
        subject: sc.dids.bob,
        includeAllUserRecords: true,
      })

      const getPaginatedEvents = async (
        sortDirection: 'asc' | 'desc' = 'desc',
      ) => {
        let defaultCursor: undefined | string = undefined
        const events: ComAtprotoAdminDefs.ModEventView[] = []
        let count = 0
        do {
          // get 1 event at a time and check we get all events
          const { data } = await queryModerationEvents({
            limit: 1,
            subject: sc.dids.bob,
            includeAllUserRecords: true,
            cursor: defaultCursor,
            sortDirection,
          })
          events.push(...data.events)
          defaultCursor = data.cursor
          count++
          // The count is a circuit breaker to prevent infinite loop in case of failing test
        } while (defaultCursor && count < 10)

        return events
      }

      const defaultEvents = await getPaginatedEvents()
      const reversedEvents = await getPaginatedEvents('asc')

      expect(allEvents.data.events.length).toEqual(4)
      expect(defaultEvents.length).toEqual(allEvents.data.events.length)
      expect(reversedEvents.length).toEqual(allEvents.data.events.length)
      expect(reversedEvents[0].id).toEqual(defaultEvents[3].id)
    })
  })

  describe('get event', () => {
    it('gets an event by specific id', async () => {
      const { data } = await pdsAgent.api.com.atproto.admin.getModerationEvent(
        {
          id: 1,
        },
        {
          headers: network.bsky.adminAuthHeaders('moderator'),
        },
      )

      expect(forSnapshot(data)).toMatchSnapshot()
    })
  })
})
