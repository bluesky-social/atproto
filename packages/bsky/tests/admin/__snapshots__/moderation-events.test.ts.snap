// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`moderation-events get event gets an event by specific id 1`] = `
Object {
  "createdAt": "1970-01-01T00:00:00.000Z",
  "createdBy": "user(2)",
  "event": Object {
    "$type": "com.atproto.admin.defs#modEventReport",
    "comment": "X",
    "reportType": "com.atproto.moderation.defs#reasonMisleading",
  },
  "id": 1,
  "subject": Object {
    "$type": "com.atproto.admin.defs#repoView",
    "did": "user(0)",
    "handle": "alice.test",
    "indexedAt": "1970-01-01T00:00:00.000Z",
    "moderation": Object {
      "subjectStatus": Object {
        "createdAt": "1970-01-01T00:00:00.000Z",
        "id": 1,
        "lastReportedAt": "1970-01-01T00:00:00.000Z",
        "lastReviewedAt": "1970-01-01T00:00:00.000Z",
        "lastReviewedBy": "user(1)",
        "reviewState": "com.atproto.admin.defs#reviewEscalated",
        "subject": Object {
          "$type": "com.atproto.admin.defs#repoRef",
          "did": "user(0)",
        },
        "subjectBlobCids": Array [],
        "subjectRepoHandle": "alice.test",
        "takendown": false,
        "updatedAt": "1970-01-01T00:00:00.000Z",
      },
    },
    "relatedRecords": Array [
      Object {
        "$type": "app.bsky.actor.profile",
        "avatar": Object {
          "$type": "blob",
          "mimeType": "image/jpeg",
          "ref": Object {
            "$link": "cids(0)",
          },
          "size": 3976,
        },
        "description": "its me!",
        "displayName": "ali",
        "labels": Object {
          "$type": "com.atproto.label.defs#selfLabels",
          "values": Array [
            Object {
              "val": "self-label-a",
            },
            Object {
              "val": "self-label-b",
            },
          ],
        },
      },
    ],
  },
  "subjectBlobCids": Array [],
  "subjectBlobs": Array [],
}
`;

exports[`moderation-events query events returns all events for record or repo 1`] = `
Array [
  Object {
    "createdAt": "1970-01-01T00:00:00.000Z",
    "createdBy": "user(1)",
    "creatorHandle": "alice.test",
    "event": Object {
      "$type": "com.atproto.admin.defs#modEventReport",
      "comment": "X",
      "reportType": "com.atproto.moderation.defs#reasonSpam",
    },
    "id": 7,
    "subject": Object {
      "$type": "com.atproto.admin.defs#repoRef",
      "did": "user(0)",
    },
    "subjectBlobCids": Array [],
    "subjectHandle": "bob.test",
  },
  Object {
    "createdAt": "1970-01-01T00:00:00.000Z",
    "createdBy": "user(1)",
    "creatorHandle": "alice.test",
    "event": Object {
      "$type": "com.atproto.admin.defs#modEventReport",
      "comment": "X",
      "reportType": "com.atproto.moderation.defs#reasonSpam",
    },
    "id": 3,
    "subject": Object {
      "$type": "com.atproto.admin.defs#repoRef",
      "did": "user(0)",
    },
    "subjectBlobCids": Array [],
    "subjectHandle": "bob.test",
  },
]
`;

exports[`moderation-events query events returns all events for record or repo 2`] = `
Array [
  Object {
    "createdAt": "1970-01-01T00:00:00.000Z",
    "createdBy": "user(0)",
    "creatorHandle": "bob.test",
    "event": Object {
      "$type": "com.atproto.admin.defs#modEventReport",
      "comment": "X",
      "reportType": "com.atproto.moderation.defs#reasonSpam",
    },
    "id": 6,
    "subject": Object {
      "$type": "com.atproto.repo.strongRef",
      "cid": "cids(0)",
      "uri": "record(0)",
    },
    "subjectBlobCids": Array [],
    "subjectHandle": "alice.test",
  },
  Object {
    "createdAt": "1970-01-01T00:00:00.000Z",
    "createdBy": "user(0)",
    "creatorHandle": "bob.test",
    "event": Object {
      "$type": "com.atproto.admin.defs#modEventReport",
      "comment": "X",
      "reportType": "com.atproto.moderation.defs#reasonSpam",
    },
    "id": 2,
    "subject": Object {
      "$type": "com.atproto.repo.strongRef",
      "cid": "cids(0)",
      "uri": "record(0)",
    },
    "subjectBlobCids": Array [],
    "subjectHandle": "alice.test",
  },
]
`;
